### PR TITLE
Mini batch training using point set boundary condition

### PR DESCRIPTION
**Pull request by mattragoza**
This pull request implements `BatchPointSetBC`, as described in #744 . This is a subclass of `PointSetBC` that takes additional constructor arguments of `batch_size` and `shuffle`. It uses a `deepxde.data.sampler.BatchSampler` to maintain the internal state while iterating over the data points. Each call to `collocation_points` advances the iterator and returns the next batch of data points, then subsequent calls to `error` compute the boundary condition residual with respect to the data values for the current batch.

I also created an example script `examples/pinn_inverse/elliptic_inverse_field_batch.py` and verified that the performance is similar to the non-minibatch version.

### CODE REVIEW

**Comment by mattragoza** [comment]
I am aware that the PR contains additional commits from my dev branch that were already merged in a previous PR. I'm not sure what the best practice is for omitting these commits from future pull requests given that this repo uses squash merging. Should I just create a separate branch for each pull request?

**Comment by lululxvi** [comment]
> I am aware that the PR contains additional commits from my dev branch that were already merged in a previous PR. I'm not sure what the best practice is for omitting these commits from future pull requests given that this repo uses squash merging. Should I just create a separate branch for each pull request?

Yes, that would be better. It is fine this time.

**Comment by mattragoza** [comment]
I have only tested with pytorch so far, but I will test with other backends.

**Comment by mattragoza** [comment]
Maybe we can add an additional method `pde.resample_bc_points` to PDE, along with a BCResidualResampler callback?

**Comment by lululxvi** [comment]
> Maybe we can add an additional method `pde.resample_bc_points` to PDE, along with a BCResidualResampler callback?

Yes, adding a callback is a good idea.

**Comment by lululxvi** [comment]
It might be better to combine `BatchPointSetBC` and `PointSetBC` into one class, by checking if `batch_size` is `None`.

File: deepxde/callbacks.py [hunk 1]
```diff
@@ -510,3 +510,12 @@ def on_epoch_end(self):
...
             )
+
+
+class BatchResampler(Callback):
+    """Resample the training points for both PDE and boundary condition losses after
+    each step."""
+
+    def on_batch_end(self):
+        self.model.data.train_x_bc = None
+        self.model.data.resample_train_points()
```
Comment (lululxvi) [inline deepxde/callbacks.py:521]: This will also resample the PDE training points. It might be better to add a function `resample_bc_points` in `dde.data.PDE`, similar to
https://github.com/lululxvi/deepxde/blob/09c74ab6a5c3269e248b09f27518a05d506b0327/deepxde/data/pde.py#L196

but probably a little more complicated .

Comment (mattragoza) [reply deepxde/callbacks.py:521]: I agree but this requires more changes to the PDE class. The approach I would take is to move the `run_if_all_none("train_x", "train_y", "train_aux_vars")` decorator from `pde.train_next_batch` onto `pde.train_points`. That way the caching of the PDE domain points and the BC points are cached separately, but `train_next_batch` runs each time and recomputes whichever points are needed.

Comment (mattragoza) [reply deepxde/callbacks.py:521]: Otherwise `pde.bc_points` won't ever get called to recompute the BC points. However this approach would also call the true solution function every time, which could be expensive. Any better ideas?

Comment (mattragoza) [reply deepxde/callbacks.py:521]: For my use case I want to resample both the PDE domain and BC points anyway.

Comment (lululxvi) [reply deepxde/callbacks.py:521]: Yes, it is a good idea to cache `pde.train_points` in `self.train_x_all`. But you can still keep `run_if_all_none("train_x", "train_y", "train_aux_vars")`.

If we resample, then re-calling the true solution function seems unavoidable.

File: deepxde/data/pde.py [hunk 2]
```diff
@@ -113,11 +113,12 @@ def __init__(
 
         self.auxiliary_var_fn = auxiliary_var_function
 
-        # TODO: train_x_all is used for PDE losses. It is better to add train_x_pde explicitly.
-        self.train_x_all = None
-        self.train_x, self.train_y = None, None
+        self.train_x_pde = None
```
Comment (lululxvi) [inline deepxde/data/pde.py:116]: Don't modify the behavior of `train_x_all`, otherwise many other classes depending on `PDE` will break.

File: deepxde/callbacks.py [hunk 3]
```diff
@@ -502,11 +502,19 @@ def on_epoch_end(self):
...
                 "`num_bcs` changed! Please update the loss function by `model.compile`."
             )
+
+
+class BatchResampler(Callback):
+    """Resample the training points for both PDE and boundary condition losses after
+    each step."""
+
+    def on_batch_end(self):
+        self.model.data.resample_train_points()
```
Comment (lululxvi) [inline deepxde/callbacks.py:520]: Resample on every itereation seems expensive. Maybe add `period` as `PDEResidualResampler`?

File: deepxde/data/pde.py (hunk 2 above)
Comment (mattragoza) [reply deepxde/data/pde.py:116]: Ok. I found the variable and method names confusing, since `train_x_all` only contains the PDE points and `train_x` contains all the training points including PDE and BC points. I would be willing to open a separate pull request that makes this change and any necessary changes to other classes that would be affected (I think just `TimePDE`, `FPDE`, and `TimeFPDE`).

File: deepxde/callbacks.py (hunk 3 above)
Comment (mattragoza) [reply deepxde/callbacks.py:520]: I will make this change.

File: deepxde/data/pde.py (hunk 2 above)
Comment (lululxvi) [reply deepxde/data/pde.py:116]: Yes, use a separate PR. There are more classes using it, such as `PDEOperator`, `PDEOperatorCartesianProd`.

File: deepxde/callbacks.py (hunk 3 above)
Comment (mattragoza) [reply deepxde/callbacks.py:520]: What do you think about combining into a single class that takes arguments for resampling the PDE points, BC points or both?

Comment (mattragoza) [reply deepxde/callbacks.py:520]: I.e. the constructor signature would be something like `BatchResampler(period=100, pde_points=True, bc_points=True)`.

Comment (lululxvi) [reply deepxde/callbacks.py:520]: Good idea. But we can do step by step. Maybe this can be another PR.

File: deepxde/callbacks.py [hunk 4]
```diff
@@ -502,11 +502,28 @@ def on_epoch_end(self):
...
 
         if not np.array_equal(self.num_bcs_initial, self.model.data.num_bcs):
             print("Initial value of self.num_bcs:", self.num_bcs_initial)
             print("self.model.data.num_bcs:", self.model.data.num_bcs)
             raise ValueError(
                 "`num_bcs` changed! Please update the loss function by `model.compile`."
             )
+
+
+class BatchResampler(Callback):
```
Comment (lululxvi) [inline deepxde/callbacks.py:515]: How about `PDEResampler`?

File: deepxde/callbacks.py (hunk 3 above)
Comment (lululxvi) [reply deepxde/callbacks.py:529]: Actually I agree with you it is good to combine them now, because there are so many same code.

File: deepxde/data/pde.py [hunk 5]
```diff
@@ -225,6 +241,7 @@ def replace_with_anchors(self, anchors):
                 config.real(np)
             )
 
+    @run_if_all_none("train_x_all")
```
Comment (lululxvi) [inline deepxde/data/pde.py:244]: Do you also need to add this in line 311?

File: deepxde/icbc/boundary_conditions.py [hunk 6]
```diff
@@ -17,6 +17,7 @@
 import numpy as np
 
 from .. import backend as bkd
+from .. import data
 from .. import config
```
Comment (lululxvi) [inline deepxde/icbc/boundary_conditions.py:21]:
```
from .. import config
from .. import data
```

File: deepxde/data/pde.py (hunk 5 above)
Comment (mattragoza) [reply deepxde/data/pde.py:244]: Yes, you're right

File: deepxde/callbacks.py [hunk 7]
```diff
@@ -484,12 +484,20 @@ def on_train_end(self):
                 )
 
 
-class PDEResidualResampler(Callback):
-    """Resample the training points for PDE losses every given period."""
+class PDEResampler(Callback):
```
Comment (lululxvi) [inline deepxde/callbacks.py:487]: How about `PDEPointResampler`? so that it is easier to understand.

File: examples/pinn_forward/heat.py [hunk 8]
```diff
@@ -1,3 +1,4 @@
+"""Backend supported: tensorflow.compat.v1, pytorch"""
```
Comment (lululxvi) [inline examples/pinn_forward/heat.py:1]: Why not "tensorflow"?

Comment (mattragoza) [reply examples/pinn_forward/heat.py:1]: I have not tested with tensorflow yet because I had trouble installing tensorflow_probability in my conda environment.

File: deepxde/icbc/boundary_conditions.py [hunk 9]
```diff
@@ -180,11 +183,31 @@ def __init__(self, points, values, component=0):
...
     def collocation_points(self, X):
+        if self.batch_size is not None:
+            self.batch_indices = self.batch_sampler.get_next(self.batch_size)
+            return self.points[self.batch_indices]
         return self.points
 
     def error(self, X, inputs, outputs, beg, end, aux_var=None):
+        if self.batch_size is not None:
+            if backend_name == "tensorflow.compat.v1":
+                batch_values = bkd.tf.gather(self.values, self.batch_indices)
```
Comment (lululxvi) [inline deepxde/icbc/boundary_conditions.py:204]: I am not sure if this would work or not, because in TF 1, `self.batch_indices` may only be initialized once in the graph.

Also, in TF 2, we use `@tf.function` for graph mode...

Comment (mattragoza) [reply deepxde/icbc/boundary_conditions.py:204]: I really am not very familiar with tensorflow. Maybe I should just keep support for pytorch for now.

File: examples/pinn_forward/heat.py (hunk 8 above)
Comment (lululxvi) [reply examples/pinn_forward/heat.py:1]: OK. I just tested it, and made a few changes, see https://github.com/lululxvi/deepxde/commit/b927780de668ccb8f5b01c15985cf04cc905665c
You can remove it from your PR.

File: deepxde/icbc/boundary_conditions.py (hunk 9 above)
Comment (lululxvi) [reply deepxde/icbc/boundary_conditions.py:204]: Yes, I agree. Check the backend; if it is not PyTorch, then throw an error.

File: deepxde/callbacks.py [hunk 10]
```diff
@@ -484,12 +484,20 @@ def on_train_end(self):
...
+    """Resample the training points for PDE and/or BC losses every given period.
 
-    def __init__(self, period=100):
+    Args:
+        period: How often to resample the training points (default is 100 iterations).
+        pde_points: If True, resample the training points for PDE losses.
+        bc_points: If True, resample the training points for BC losses.
+    """
+
+    def __init__(self, period=100, pde_points=True, bc_points=True):
```
Comment (lululxvi) [inline deepxde/callbacks.py:496]: Use `bc_points=False` as default, as it only supports PyTorch.

File: deepxde/icbc/boundary_conditions.py [hunk 11]
```diff
@@ -169,9 +170,11 @@ class PointSetBC:
             used for training.
         values: An array of values that gives the exact solution of the problem.
         component: The output component satisfying this BC.
+        batch_size: The number of points per minibatch, or `None` to return all points.
```
Comment (lululxvi) [inline deepxde/icbc/boundary_conditions.py:173]:
```
batch_size: The number of points per minibatch, or `None` to return all points. This is only supported for the backend PyTorch.
```

File: examples/pinn_forward/heat_resample.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
+"""Backend supported: tensorflow.compat.v1, pytorch"""
+import deepxde as dde
+import matplotlib.pyplot as plt
```
Comment (lululxvi) [inline examples/pinn_forward/heat_resample.py:3]: Delete `import matplotlib.pyplot as plt`

File: deepxde/callbacks.py (hunk 10 above)
Comment (lululxvi) [reply deepxde/callbacks.py:496]: @mattragoza

Comment (lululxvi) [reply deepxde/callbacks.py:496]: @mattragoza Would you like to finish the step of this PR?

Comment (mattragoza) [reply deepxde/callbacks.py]: Done
